stdlib/public/Python/PythonLibrary.swift

```diff
@@ -0,0 +1,184 @@
+  ) -> T {
+    var name = name
+    if let legacyName = legacyName, PythonLibrary.shared.isLegacyPython {
+      name = legacyName
+    }
+    
+    log(message: "Loading symbol '\(name)' from the Python library...")
+    return unsafeBitCast(
+      dlsym(PythonLibrary.shared.pythonLibraryHandle, name),
+      to: signature
+    )
+  }
+}
+
+// Methods of `PythonLibrary` required to load the Python library.
+extension PythonLibrary {
+  private static let supportedMajorVersions: [Int] =
+    Array(2...3).reversed()
+  private static let supportedMinorVersions: [Int?] =
+    [nil] + Array(0...30).reversed()
+  
+  private static let libraryPathVersionCharacter: Character = ":"
+  
+  #if canImport(Darwin)
+  private static var libraryNames = ["Python.framework/Versions/:/Python"]
+  private static var libraryPathExtensions = [""]
+  private static var librarySearchPaths = ["", "/usr/local/Frameworks/"]
+  #elseif os(Linux)
+  private static var libraryNames = ["libpython:", "libpython:m"]
+  private static var libraryPathExtensions = [".so"]
+  private static var librarySearchPaths = [""]
+  #endif
+  
+  private static let libraryPaths: [String] = {
+    var libraryPaths: [String] = []
+    for librarySearchPath in librarySearchPaths {
+      for libraryName in libraryNames {
+        for libraryPathExtension in libraryPathExtensions {
+          let libraryPath =
+            librarySearchPath + libraryName + libraryPathExtension
+          libraryPaths.append(libraryPath)
+        }
+      }
+    }
+    
+    return libraryPaths
+  }()
+  
+  private static func loadPythonLibrary(
+    at path: String, majorVersion: Int, minorVersion: Int? = nil
+  ) -> UnsafeMutableRawPointer? {
+    var versionString = String(majorVersion)
+    if let minorVersion = minorVersion {
+      versionString += ".\(minorVersion)"
+    }
+    
+    if let requiredPythonVersion = Environment.version.value {
+      let requiredMajorVersion = Int(requiredPythonVersion)
+      if requiredPythonVersion != versionString,
+        requiredMajorVersion != majorVersion {
+        return nil
+      }
+    }
+    
+    let path = path.split(separator: libraryPathVersionCharacter)
+      .joined(separator: versionString)
+    return loadPythonLibrary(at: path)
+  }
+  
+  private static func loadPythonLibrary(
+    at path: String
+  ) -> UnsafeMutableRawPointer? {
+    log(message: "Trying to load library at '\(path)'...")
+    let pythonLibraryHandle = dlopen(path, RTLD_LAZY)
+    
+    if pythonLibraryHandle != nil {
+      log(message: "Library at '\(path)' was sucessfully loaded.")
+    }
+    return pythonLibraryHandle
+  }
+  
+  private static func getPythonLibraryHandle() -> UnsafeMutableRawPointer? {
+    if let pythonLibraryPath = Environment.library.value {
+      return loadPythonLibrary(at: pythonLibraryPath)
+    }
+    
+    for majorVersion in supportedMajorVersions {
+      for minorVersion in supportedMinorVersions {
```

Comment:
Would this be more elegant if there were simply an array of tuples called `supportedVersions: [(major: Int, minor: Int?)]`?

Such a tuple could also be manually populated (i.e., `[(3, nil), (3, 8), (3, 7), /* ... */]`). There's no need for supporting nonexistent Python versions (indeed, that would be an oxymoron).

Reply:
Removing support for "nonexistent" Python versions means that old Swift distributions (or even any distributed executable that statically linked the Python module) would not work with any 3.x Python version released after it, even though the Python ABI is marked as stable for all the Python 3 lifetime. Also, it would require maintaining the list and deciding when it's time to add new versions.

Take in account that while in Windows the Python distribution includes both `python3.dll` and `python37.dll` in macOS and Linux we have to specify the exact minor version to load it (`Python.framework/Versions/3.7/Python` or `libpython3.7m.so`).

Reply:
If the ABI is stable for all versions of Python 3, then is there a need for a list of supported minor versions?

Reply:
@xwu Read the second paragraph.

Reply:
I'm afraid I don't understand. Why do we need a list of supported minor versions?

Reply:
I understand that. What I don't understand is why we need a static array of "supported" minor versions if support is not contingent on the minor version.

Reply:
We can work on this in a follow-up.

There will never be a version of Python 2 after 2.7 for instance, and it makes no sense to look for one; if "Python 2.30" existed, we certainly wouldn't want to support it. Likewise, is Python 2.0 really supported? This is all to say that iterating over the same minor versions for every major version is not ideal.